### PR TITLE
Make `Util.memoize` thread-safe

### LINKED ISSUE #9151: RenderType memoization is not thread safe
This was originally reported to me for 1.18 (see https://github.com/cc-tweaked/CC-Tweaked/issues/1213), but the underlying code is still the same on 1.19.2 

**Minecraft Version:** 1.19.2
**Forge Version:** 43.1.52
**Steps to Reproduce:**
 1. Create multiple mods which call `RenderType.text(...)` during mod construction (or any of the other multi-threaded initialisation stages).
 2. Start the game.
 3. The game crashes with a `ConcurrentModificationException` inside `Util.memoize`

**Description of issue:**
Minecraft, and several mods, store `RenderType`s within `static final` fields. However, if this render type is created using one of the smart constructors (such as `RenderType.text` or `RenderType.entityNoOutline`), it is first looked up in a cache, and then added if not present.

However, as mod construction and initialisation occurs in parallel, multiple mods may attempt to call one of the smart constructors at once. This causes a `ConcurrentModificationException` in the underlying cache.

Given [the fairly widespread nature](https://github.com/search?q=language%3Ajava+%22static+final+RenderType%22&type=code) of this usage of RenderTypes, and the unpredictability of class loading, it feels this is something best patched in Forge. The obvious solution would be to switch from using a `HashMap` to a `ConcurrentHashMap`. I'm happy to put a PR together for this, if that sounds sensible?

### PR DESCRIPTION
**Pull request by SquidDev**
This switches `Util.memoize` from using a `HashMap` to a `ConcurrentHashMap`, ensuring it is safe to access during multi-threaded mod loading. Fixes #9151.

### CODE REVIEW

**Review by marchermans — APPROVED** [review]

**Comment by tmvkrpxl0** [comment]
Would this have any measurable impact on performance?

**Comment by SquidDev** [comment]
> Would this have any measurable impact on performance?

**TLDR:** No. On the hot path (i.e. the computed value is already present), `ConcurrentHashMap.computeIfAbsent` is lock free. The performance difference between `HashMap` and `ConcurrentHashMap`'s implementation is negligible.

---

I put together a small JMH benchmark for `computeIfAbsent`.

<details><summary><code>MapBenchmark.java</code></summary>

```java
import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.Scope;
import org.openjdk.jmh.annotations.State;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.RunnerException;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;

public class MapBenchmark {
  private static final Function<ResourceLocation, ResourceLocation> COMPUTE = x -> x;
  private static final ResourceLocation OUR_ID = new ResourceLocation("our_mod", "our_path");

  public static class MapHolder {
    final Map<ResourceLocation, ResourceLocation> map;

    MapHolder(Map<ResourceLocation, ResourceLocation> map) {
      this.map = map;

      // Populate the cache with a small number of items.
      for (int i = 0; i < 10; i++) {
        map.computeIfAbsent(new ResourceLocation("mod_" + i, "path_" + i), COMPUTE);
      }
      map.computeIfAbsent(OUR_ID, COMPUTE);
    }
  }

  @State(Scope.Benchmark)
  public static class HashMapHolder extends MapHolder {
    public HashMapHolder() {
      super(new HashMap<>());
    }
  }

  @State(Scope.Benchmark)
  public static class ConcurrentHashMapHolder extends MapHolder {
    public ConcurrentHashMapHolder() {
      super(new ConcurrentHashMap<>());
    }
  }

  @Benchmark
  public void hashMap(HashMapHolder holder) {
    holder.map.computeIfAbsent(OUR_ID, COMPUTE);
  }

  @Benchmark
  public void concurrentHashMap(ConcurrentHashMapHolder holder) {
    holder.map.computeIfAbsent(OUR_ID, COMPUTE);
  }

  public static void main(String... args) throws RunnerException {
    Options opts = new OptionsBuilder().include(MapBenchmark.class.getName() + ".*")
        .forks(2)
        .build();
    new Runner(opts).run();
  }

  private static class ResourceLocation {
    private final String namespace;
    private final String path;

    private ResourceLocation(String namespace, String path) {
      this.namespace = namespace;
      this.path = path;
    }

    public boolean equals(Object other) {
      if (this == other) return true;
      if (!(other instanceof ResourceLocation)) return false;

      ResourceLocation otherResource = (ResourceLocation) other;
      return namespace.equals(otherResource.namespace) && path.equals(otherResource.path);
    }

    public int hashCode() {
      return 31 * namespace.hashCode() + path.hashCode();
    }
  }
}
```

</details>

Under Java 17 (JDK 17.0.2, OpenJDK 64-Bit Server VM, 17.0.2+8, Linux), I get the following, suggesting `ConcurrentHashMap` is (slightly) faster than `HashMap`.

```
Benchmark                        Mode  Cnt          Score         Error  Units
MapBenchmark.concurrentHashMap  thrpt   10  194983532.829 ± 1315109.211  ops/s
MapBenchmark.hashMap            thrpt   10  157983178.105 ± 1268198.718  ops/s
```

This isn't the case on Java 8, where `HashMap` is much faster (it reaches 250M ops/s). I'm not sure why - there's been a few minor changes to `HashMap.computeIfAbsent` to correctly check for concurrent modifications, but surprised it makes this much of a difference. Afraid I haven't got round to digging out jitwatch :p.

Either way, each `computeIfAbsent` call is taking at most 6ns. This is pretty insignificant compared with pretty most other rendering code (i.e. `MultiBufferSource.BufferSource.getBuffer`).

**Comment by tmvkrpxl0** [comment]
well it was more of just curiosity than a claim but thanks for testing it anyway :smile:

**Review by Technici4n — APPROVED** [review]
I was concerned by the performance, good that you tested it.

**Comment by SquidDev** [comment]
Thanks!
